File: .zo/scripts/python/check-prerequisites.py
#!/usr/bin/env python3
"""
Consolidated prerequisite checking script for Spec-Driven Development workflow.

This script provides unified prerequisite checking for Spec-Driven Development workflow.
It replaces the functionality previously spread across multiple bash scripts.

This script uses the shared `common.py` module for core utility functions.

Usage: python check-prerequisites.py [OPTIONS]

OPTIONS:
  --json              Output in JSON format
  --require-tasks     Require tasks.md to exist (for implementation phase)
  --include-tasks     Include tasks.md in AVAILABLE_DOCS list
  --paths-only        Only output path variables (no validation)
  --help, -h          Show help message

OUTPUTS:
  JSON mode: {"FEATURE_DIR":"...", "AVAILABLE_DOCS":["..."]}
  Text mode: FEATURE_DIR:... \n AVAILABLE_DOCS: \n ✓/✗ file.md
  Paths only: REPO_ROOT: ... \n BRANCH: ... \n FEATURE_DIR: ... etc.
"""

import argparse
import json
import logging
import os
import sys

# Import shared utility functions from common.py
from common import (
    get_feature_paths,
    has_git,
    get_repo_root,
    get_current_branch,
    check_feature_branch,
    check_file_exists,
    check_dir_exists_with_files,
)

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(levelname)s: %(message)s'
)
logger = logging.getLogger(__name__)


# ============================================================================
# Output Functions (script-specific)
# ============================================================================

def format_json_paths(paths: dict) -> str:
    """
    Format paths as JSON for paths-only mode.
    
    Args:
        paths: Dictionary of path variables
        
    Returns:
        JSON formatted string (compact, no indentation)
    """
    return json.dumps({
        'REPO_ROOT': paths['REPO_ROOT'],
        'BRANCH': paths['CURRENT_BRANCH'],
        'FEATURE_DIR': paths['FEATURE_DIR'],
        'FEATURE_SPEC': paths['FEATURE_SPEC'],
        'IMPL_PLAN': paths['IMPL_PLAN'],
        'TASKS': paths['TASKS'],
    }, separators=(',', ':'))


def format_json_result(feature_dir: str, available_docs: list) -> str:
    """
    Format result as JSON.
    
    Args:
        feature_dir: Feature directory path
        available_docs: List of available document names
        
    Returns:
        JSON formatted string (compact)
    """
    result = {
        'FEATURE_DIR': feature_dir,
        'AVAILABLE_DOCS': available_docs,
    }
    return json.dumps(result, separators=(',', ':'))


def check_file_status(file_path: str, display_name: str) -> str:
    """
    Check file status and return formatted string.
    
    Args:
        file_path: Path to file
        display_name: Name to display
        
    Returns:
        Formatted status string with checkmark or cross
    """
    if check_file_exists(file_path):
        return f"  ✓ {display_name}"
    else:
        return f"  ✗ {display_name}"


def check_dir_status(dir_path: str, display_name: str) -> str:
    """
    Check directory status and return formatted string.
    
    Args:
        dir_path: Path to directory
        display_name: Name to display
        
    Returns:
        Formatted status string with checkmark or cross
    """
    if check_dir_exists_with_files(dir_path):
        return f"  ✓ {display_name}"
    else:
        return f"  ✗ {display_name}"


def print_help():
    """Print help message."""
    help_text = """Usage: check-prerequisites.sh [OPTIONS]

Consolidated prerequisite checking for Spec-Driven Development workflow.

OPTIONS:
  --json              Output in JSON format
  --require-tasks     Require tasks.md to exist (for implementation phase)
  --include-tasks     Include tasks.md in AVAILABLE_DOCS list
  --paths-only        Only output path variables (no prerequisite validation)
  --help, -h          Show this help message

EXAMPLES:
  # Check task prerequisites (plan.md required)
  ./check-prerequisites.sh --json
  
  # Check implementation prerequisites (plan.md + tasks.md required)
  ./check-prerequisites.sh --json --require-tasks --include-tasks
  
  # Get feature paths only (no validation)
  ./check-prerequisites.sh --paths-only
  
"""
    print(help_text)


# ============================================================================
# Main Logic
# ============================================================================

def main():
    """Main entry point for the script."""
    # Parse command line arguments
    parser = argparse.ArgumentParser(add_help=False)
    parser.add_argument('--json', action='store_true', help='Output in JSON format')
    parser.add_argument('--require-tasks', action='store_true', help='Require tasks.md to exist')
    parser.add_argument('--include-tasks', action='store_true', help='Include tasks.md in AVAILABLE_DOCS list')
    parser.add_argument('--paths-only', action='store_true', help='Only output path variables (no validation)')
    parser.add_argument('--help', '-h', action='store_true', help='Show help message')
    
    # Check for unknown arguments
    args, unknown = parser.parse_known_args()
    if unknown:
        print(f"ERROR: Unknown option '{unknown[0]}'. Use --help for usage information.", file=sys.stderr)
        sys.exit(1)
    
    # Handle help flag
    if args.help:
        print_help()
        sys.exit(0)
    
    logger.debug("Starting prerequisite check")
    
    # Get feature paths and validate branch
    paths = get_feature_paths()
    logger.debug(f"Feature paths: {paths}")
    
    is_valid, error_msg = check_feature_branch(
        paths['CURRENT_BRANCH'], 
        paths['HAS_GIT'] == 'true'
    )
    
    if not is_valid:
        print(f"ERROR: {error_msg}", file=sys.stderr)
        sys.exit(1)
    
    # If paths-only mode, output paths and exit
    if args.paths_only:
        if args.json:
            print(format_json_paths(paths))
        else:
            print(f"REPO_ROOT: {paths['REPO_ROOT']}")
            print(f"BRANCH: {paths['CURRENT_BRANCH']}")
            print(f"FEATURE_DIR: {paths['FEATURE_DIR']}")
            print(f"FEATURE_SPEC: {paths['FEATURE_SPEC']}")
            print(f"IMPL_PLAN: {paths['IMPL_PLAN']}")
            print(f"TASKS: {paths['TASKS']}")
        sys.exit(0)
    
    # Validate required directories and files
    if not os.path.isdir(paths['FEATURE_DIR']):
        print(f"ERROR: Feature directory not found: {paths['FEATURE_DIR']}", file=sys.stderr)
        print("Run /zo.zo first to create the feature structure.", file=sys.stderr)
        sys.exit(1)
    
    if not os.path.isfile(paths['IMPL_PLAN']):
        print(f"ERROR: plan.md not found in {paths['FEATURE_DIR']}", file=sys.stderr)
        print("Run /zo.plan first to create the implementation plan.", file=sys.stderr)
        sys.exit(1)
    
    # Check for tasks.md if required
    if args.require_tasks and not os.path.isfile(paths['TASKS']):
        print(f"ERROR: tasks.md not found in {paths['FEATURE_DIR']}", file=sys.stderr)
        print("Run /zo.tasks first to create the task list.", file=sys.stderr)
        sys.exit(1)
    
    # Build list of available documents
    available_docs = []
    
    # Always check these optional docs
    if check_file_exists(paths['RESEARCH']):
        available_docs.append('research.md')
    if check_file_exists(paths['DATA_MODEL']):
        available_docs.append('data-model.md')
    if check_file_exists(paths['DESIGN_FILE']):
        available_docs.append('design.md')
    
    # Check contracts directory (only if it exists and has files)
    if check_dir_exists_with_files(paths['CONTRACTS_DIR']):
        available_docs.append('contracts/')
    
    if check_file_exists(paths['QUICKSTART']):
        available_docs.append('quickstart.md')
    
    # Include tasks.md if requested and it exists
    if args.include_tasks and check_file_exists(paths['TASKS']):
        available_docs.append('tasks.md')
    
    # Output results
    if args.json:
        print(format_json_result(paths['FEATURE_DIR'], available_docs))
    else:
        # Text output
        print(f"FEATURE_DIR:{paths['FEATURE_DIR']}")
        print("AVAILABLE_DOCS:")
        
        # Show status of each potential document
        print(check_file_status(paths['RESEARCH'], 'research.md'))
        print(check_file_status(paths['DATA_MODEL'], 'data-model.md'))
        print(check_file_status(paths['DESIGN_FILE'], 'design.md'))
        print(check_dir_status(paths['CONTRACTS_DIR'], 'contracts/'))
        print(check_file_status(paths['QUICKSTART'], 'quickstart.md'))
        
        if args.include_tasks:
            print(check_file_status(paths['TASKS'], 'tasks.md'))
    
    logger.debug("Prerequisite check completed successfully")
    sys.exit(0)


if __name__ == '__main__':
    main()

File: .roo/tools/check-prerequisites.ts
/**
 * check_prerequisites - Validate feature prerequisites and available documentation
 *
 * Checks if the current feature has required documentation files and directories.
 * Returns available optional docs (research.md, data-model.md, design.md, contracts/, quickstart.md).
 * Validates prerequisites for Spec-Driven Development workflow operations.
 *
 * Ported from: .zo/scripts/python/check-prerequisites.py
 *
 * @example
 * ```typescript
 * const result = await check_prerequisites({ json: true })
 * // Returns: {"FEATURE_DIR": "/path/to/specs/001-feature", "AVAILABLE_DOCS": ["research.md"]}
 * ```
 *
 * @param json - Output in JSON format (default: false)
 * @param requireTasks - Require tasks.md to exist (for implementation phase, default: false)
 * @param includeTasks - Include tasks.md in AVAILABLE_DOCS list (default: false)
 * @param pathsOnly - Only output path variables, no validation (default: false)
 *
 * @returns JSON object or formatted text with feature paths and available docs
 *
 * @throws {Error} If execution environment validation fails
 * @throws {Error} If not on a valid feature branch
 * @throws {Error} If feature directory not found
 * @throws {Error} If plan.md not found
 * @throws {Error} If tasks.md required but not found
 *
 * @see {@link ../lib/util.ts} for path resolution utilities
 */

import { defineCustomTool } from '@roo-code/types';
import { parametersSchema as z } from '@roo-code/types';
import { existsSync } from 'node:fs';
import { join } from 'node:path';
import {
  checkDirExistsWithFiles,
  checkFeatureBranch,
  checkFileExists,
  getFeaturePaths,
  getRepoRoot,
  resolvePath,
  validateExecutionEnvironment,
} from './lib/util.js';

export default defineCustomTool({
  name: 'check_prerequisites',

  description: `
    Check if the current feature branch has required documentation files.
    Returns available optional docs (research.md, data-model.md, design.md, contracts/, quickstart.md).
    Validates prerequisites for Spec-Driven Development workflow operations.
  `,

  parameters: z.object({
    json: z
      .boolean()
      .optional()
      .describe('Output results in JSON format (compact, no indentation)'),
    requireTasks: z
      .boolean()
      .optional()
      .describe('Require tasks.md to exist (for implementation phase)'),
    includeTasks: z
      .boolean()
      .optional()
      .describe('Include tasks.md in AVAILABLE_DOCS list'),
    pathsOnly: z
      .boolean()
      .optional()
      .describe('Only output path variables without validation'),
  }),

  async execute({ json = false, requireTasks = false, includeTasks = false, pathsOnly = false }, _context) {
    try {
      // Validate execution environment first
      if (!validateExecutionEnvironment()) {
        const errorMsg = 'ERROR: Execution environment validation failed.';
        return json
          ? JSON.stringify({ error: errorMsg, valid: false })
          : errorMsg;
      }

      // Get feature paths and validate branch
      const paths = getFeaturePaths();

      // Check if on valid feature branch
      const branchValidation = checkFeatureBranch(paths.CURRENT_BRANCH, paths.HAS_GIT);
      if (!branchValidation.isValid) {
        const errorMsg = `ERROR: ${branchValidation.error}`;
        return json
          ? JSON.stringify({ error: errorMsg, valid: false })
          : errorMsg;
      }

      // If paths-only mode, output paths and exit
      if (pathsOnly) {
        const pathsData = {
          REPO_ROOT: paths.REPO_ROOT,
          BRANCH: paths.CURRENT_BRANCH,
          FEATURE_DIR: paths.FEATURE_DIR,
          FEATURE_SPEC: paths.FEATURE_SPEC,
          IMPL_PLAN: paths.IMPL_PLAN,
          TASKS: paths.TASKS,
        };
        return json
          ? JSON.stringify(pathsData)
          : [
              `REPO_ROOT: ${pathsData.REPO_ROOT}`,
              `BRANCH: ${pathsData.BRANCH}`,
              `FEATURE_DIR: ${pathsData.FEATURE_DIR}`,
              `FEATURE_SPEC: ${pathsData.FEATURE_SPEC}`,
              `IMPL_PLAN: ${pathsData.IMPL_PLAN}`,
              `TASKS: ${pathsData.TASKS}`,
            ].join('\n');
      }

      // Resolve paths for file operations
      const featureDir = resolvePath(paths.FEATURE_DIR);
      const implPlan = resolvePath(paths.IMPL_PLAN);
      const tasks = resolvePath(paths.TASKS);
      const research = resolvePath(paths.RESEARCH);
      const dataModel = resolvePath(paths.DATA_MODEL);
      const designFile = resolvePath(paths.DESIGN_FILE);
      const contractsDir = resolvePath(paths.CONTRACTS_DIR);
      const quickstart = resolvePath(paths.QUICKSTART);

      // Validate required directories and files
      if (!existsSync(featureDir)) {
        const errorMsg = [
          `ERROR: Feature directory not found: ${featureDir}`,
          `  Current working directory: ${process.cwd()}`,
          `  Expected feature directory: ${featureDir}`,
          "Run 'zo.spec' first to create the feature structure.",
        ].join('\n');
        return json
          ? JSON.stringify({ error: errorMsg, valid: false })
          : errorMsg;
      }

      if (!existsSync(implPlan)) {
        const errorMsg = [
          `ERROR: plan.md not found in ${featureDir}`,
          "Run 'zo.plan' first to create the implementation plan.",
        ].join('\n');
        return json
          ? JSON.stringify({ error: errorMsg, valid: false })
          : errorMsg;
      }

      // Check for tasks.md if required
      if (requireTasks && !existsSync(tasks)) {
        const errorMsg = [
          `ERROR: tasks.md not found in ${featureDir}`,
          "Run 'zo.tasks' first to create the task list.",
        ].join('\n');
        return json
          ? JSON.stringify({ error: errorMsg, valid: false })
          : errorMsg;
      }

      // Build list of available documents
      const availableDocs: string[] = [];

      // Check optional docs
      if (checkFileExists(research)) {
        availableDocs.push('research.md');
      }
      if (checkFileExists(dataModel)) {
        availableDocs.push('data-model.md');
      }
      if (checkFileExists(designFile)) {
        availableDocs.push('design.md');
      }
      if (checkDirExistsWithFiles(contractsDir)) {
        availableDocs.push('contracts/');
      }
      if (checkFileExists(quickstart)) {
        availableDocs.push('quickstart.md');
      }

      // Include tasks.md if requested and it exists
      if (includeTasks && checkFileExists(tasks)) {
        availableDocs.push('tasks.md');
      }

      // Output results
      if (json) {
        const result = {
          FEATURE_DIR: paths.FEATURE_DIR,
          AVAILABLE_DOCS: availableDocs,
          HAS_TASKS: checkFileExists(tasks),
          VALID: true,
        };
        return JSON.stringify(result);
      }

      // Text output
      const lines = [
        `FEATURE_DIR: ${paths.FEATURE_DIR}`,
        'AVAILABLE_DOCS:',
        checkFile(research, 'research.md'),
        checkFile(dataModel, 'data-model.md'),
        checkFile(designFile, 'design.md'),
        checkDir(contractsDir, 'contracts/'),
        checkFile(quickstart, 'quickstart.md'),
      ];

      if (includeTasks) {
        lines.push(checkFile(tasks, 'tasks.md'));
      }

      return lines.join('\n');

    } catch (error) {
      const errorMessage = error instanceof Error ? error.message : String(error);
      return json
        ? JSON.stringify({ error: errorMessage, valid: false })
        : `ERROR: ${errorMessage}`;
    }
  },
});

/**
 * Check file status and return formatted string
 */
function checkFile(filePath: string, displayName: string): string {
  return checkFileExists(filePath) ? `  ✓ ${displayName}` : `  ✗ ${displayName}`;
}

/**
 * Check directory status and return formatted string
 */
function checkDir(dirPath: string, displayName: string): string {
  return checkDirExistsWithFiles(dirPath) ? `  ✓ ${displayName}` : `  ✗ ${displayName}`;
}
